### ISSUE TITLE
BUG : Self Hosted Rudderstack Response 400 and Error Invalid JSON

### ISSUE THREAD

**Issue by chandrabezzo** [rudderlabs/rudder-sdk-flutter#116]
**Describe the bug**
We have self hosted rudderstack in our company. Then, set flutter plugin to the latest version on `rudder_sdk_flutter: ^2.4.1`. I got an error
```
E/RudderSDK( 8304): Error: RudderNetworkManager: sendNetworkRequest: Request to endpoint {YOUR_DATA_PLANE_URL} failed with status code 400 and error Invalid JSON
```
But after some research this issue solved when you use this specific version
```
  rudder_plugin_android: 2.2.2
  rudder_plugin_ios: 2.2.2
  rudder_plugin_web: 2.2.3
  rudder_sdk_flutter_platform_interface: 2.3.1
```

**Which platform is the issue occurring on**
Is the error occurring on:

- [ ] Android
- [ ] iOS
- [x] Both Android and iOS

**Expected behavior**
Tracked an event on live events

**To Reproduce**
Running the app with active/enable write key

**Screenshots**

**Version of the _Flutter_ SDK**
[✓] Flutter (Channel stable, 3.10.6, on macOS 13.4.1 22F770820d darwin-arm64, locale en-ID)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.1)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.81.1)
[✓] Connected device (3 available)
[✓] Network resources

**SDK initialisation snippet**
Follow documentation

**Framework version of the SDK**
Flutter 3.10.6

**Check for Correct Usage of _writeKey_ and _dataPlaneUrl_**
Used active `writeKey` and correct `dataPlaneUrl`

**Comment by desusai7** [comment]
Hey @chandrabezzo, 

Can you please share the version of Rudder Server you've hosted yourself ?

**Comment by desusai7** [comment]
Can you also share detailed logs from both platforms as well ?

**Comment by chandrabezzo** [comment]
@desusai7 thank for your response. This issue was solved when we use these versions
```
  rudder_plugin_android: 2.2.2
  rudder_plugin_ios: 2.2.2
  rudder_plugin_web: 2.2.3
  rudder_sdk_flutter_platform_interface: 2.3.1
```

Maybe this information can help you? Or have you test on the latest version?
```
E/RudderSDK( 8304): Error: RudderNetworkManager: sendNetworkRequest: Request to endpoint {YOUR_DATA_PLANE_URL} failed with status code 400 and error Invalid JSON
```

In parallel I will confirm to our devops team

**Comment by chandrabezzo** [comment]
Hi @desusai7 we use version `1.3.2`

**Comment by chandrabezzo** [comment]
We need to upgrade to the latest version?

**Comment by desusai7** [comment]
Can you please upgrade to the latest version of the Rudder-Server and check once with the latest version of Flutter SDK ?

I couldn't confirm the actual reason because you've not shared the entire logs, but I am assuming the issue is because the latest version of the SDK is sending gzip encoded requests to the Rudder Server and the version of Rudder Server you've hosted is a really old one and it doesn't accept gzip encoded requests.

Updating your Rudder Server to latest one should fix this issue.

**Comment by chandrabezzo** [comment]
So @desusai7 it's a breaking changes? Maybe you can inform in the documentation about the recommendation version of the Flutter, Android/iOS SDK and the Rudderstack self hosted. Thanks.

**Comment by desusai7** [comment]
Sure @chandrabezzo, we will have this documented for all the client-side SDKs.  However, we recommend you update your self-hosted Rudderstack frequently as your current version is more than 6 months old. 

Can you please confirm if upgrading to the latest version fixed your issue ?

**Comment by chandrabezzo** [comment]
Hi @desusai7 we will try that solution after coordination with our devops team. Thank for your suggestion.

**Comment by chandrabezzo** [comment]
Hi @desusai7 this my log
```
SStack:Warn:RSNetwork: init: unable to retrieve carrier name
RSStack:Warn:RSNetwork: init: cannot retrieve carrier names on iOS 16 and above as CTCarrier is deprecated
RSStack:Warn:RSNetwork: init: unable to retrieve carrier name
RSStack:Info:RSCloudModeManager: CloudModeProcessor: EventCount: 12
RSStack:Error:RSNetworkManager: sendNetworkRequest: Request to url https://rudderstack.evermosa2z.com/v1/batch failed with statusCode 400 due to Invalid JSON
RSStack:Info:RSCloudModeManager: CloudModeProcessor: EventCount: 12
RSStack:Error:RSNetworkManager: sendNetworkRequest: Request to url https://rudderstack.evermosa2z.com/v1/batch failed with statusCode 400 due to Invalid JSON
RSStack:Info:RSCloudModeManager: CloudModeProcessor: EventCount: 12
RSStack:Error:RSNetworkManager: sendNetworkRequest: Request to url https://rudderstack.evermosa2z.com/v1/batch failed with statusCode 400 due to Invalid JSON
```

In other case, we has been set the version but still get an issue
```yaml
# Need to override related to the SDK Issue
  rudder_plugin_android: 2.2.2
  rudder_plugin_ios: 2.2.2
  rudder_plugin_web: 2.2.3
  rudder_sdk_flutter_platform_interface: 2.3.1
  rudder_sdk_flutter: 2.3.3
```

**Comment by desusai7** [comment]
Hi @chandrabezzo, 

Are you seeing this error even after upgrading to the latest version of the Rudder-Server ?

**Comment by chandrabezzo** [comment]
No @desusai7 I still use `1.3.2` version, as you can see, I still override flutter rudderstack plugin version to 2.0.0 - 2.3.0

**Comment by chandrabezzo** [comment]
I think when we use previous version from flutter rudderstack plugin will impacted to platform rudderstack (android/ios) version. So, we can still self hosted with version `1.3.2`.

**Comment by desusai7** [comment]
@chandrabezzo, Were you seeing the issue even on the previous version of the Rudderstack Flutter SDK ?

**Comment by chandrabezzo** [comment]
No, I never see this issue in the previous version.

**Comment by desusai7** [comment]
So, are you planning to stick to previous version of Rudderstack Flutter SDK to solve this issue ?

**Comment by chandrabezzo** [comment]
Yes, I think every update doesn't mean we have to keep up with the latest to maintain stability. But, if you highly recommend to upgrade to the latest, I can conclude that the supported SDK must always be updated and does not pay attention to backward compatibility.

**Comment by desusai7** [comment]
I would recommend updating at least Rudder Server to its latest version as the version you are currently on was more than 6 months old.  

Apologies for the inconvenience caused regarding this, we weren't aware of this issue so far, as all the Rudderstack maintained deployments were regularly upgraded to the latest Rudder Server Versions and we've missed testing with self-hosted ones. We always make sure to properly enforce backward compatibility and we will keep doing that.

**Comment by chandrabezzo** [comment]
Hi @desusai7 can you recommend the version about flutter plugin and self hosted? So, we can select the right version.

**Comment by ariretiarno** [comment]
hi @desusai7, any articles for upgrading rudderstack?
is it just simply change docker image version?

**Comment by desusai7** [comment]
hey @ariretiarno, 

yes that should do the work

**Comment by chandrabezzo** [comment]
Hi @andreasgangso have you replicated the issue?

**Comment by chandrabezzo** [comment]
@desusai7 thanks for your support. For now Android has been successfully tracked to the live tracking after upgrade to the latest version of Flutter SDK and Self Hosted Rudder. But, I got another issue on iOS. I didn't see any log error/issue when running iOS Flutter on simulator. Event sent without any catch error on Flutter side or call plugin method. But, I never see any log tracked on Live Event at dashboard.

**Comment by andreasgangso** [comment]
I don't think the problem is in the flutter sdk, but that it started happening at some version of the iOS pod.

Some findings:

- I tried using the latest versions (as mentioned in my last comment), and it spams 400 errors as it's trying to sync up old messages to /batch.
- I tried running with version 1.8.0 of the "Rudder" iOS pod, and it's successfully sending the events (statuscode 200 and the dbrecord count is going down)
- I also tried the latest versions but with a cleared db (i.e. I deleted the app first). Same.

**Comment by andreasgangso** [comment]
After debugging with Desu:
- The root cause seems to be that the sdk is sending the payloads gzipped, and old versions of rudder-server is rejecting that.
-  it turns out we actually did **not** properly upgrade to the newest rudder-server, so I'm waiting on my colleague to do that. You can confirm you are on the correct version by visiting <your-selfhosted-url>/version. Will update this when I've confirmed that upgrading fixes this.
- The excessive retrying is due to a bug in the iOS sdk. I will open an issue on the iOS repo.

**Comment by chandrabezzo** [comment]
@andreasgangso our team was upgraded to the self-hosted rudderstack Rudder-Server: [v1.12.1](https://github.com/rudderlabs/rudder-server/releases/tag/v1.12.1) as suggested by @desusai7, but still issue.

**Comment by andreasgangso** [comment]
> @andreasgangso our team was upgraded to the self-hosted rudderstack Rudder-Server: [v1.12.1](https://github.com/rudderlabs/rudder-server/releases/tag/v1.12.1) as suggested by @desusai7, but still issue.

But you have a new issue now, because the error 400 was fixed, correct?
So you should probably open a new github issue.

**Comment by chandrabezzo** [comment]
Yes @andreasgangso ios still untracked. But, in the old version I got same issue. I hope when upgraded version that issue will be fixed. But, unfortunately this issue still happen.

**Comment by desusai7** [comment]
Hey @chandrabezzo, 

Can you please connect over slack to debug this issue faster ?

**Comment by chandrabezzo** [comment]
@desusai7 please mention me on slack

**Comment by desusai7** [comment]
I've mentioned you on slack few hours ago and did it again now

**Comment by chandrabezzo** [comment]
@andreasgangso I got Rudder Version 1.19.1. How to get your version?

**Comment by chandrabezzo** [comment]
On iOS I never got any logs from Rudderstack SDK 😢

**Comment by desusai7** [comment]
Generally, logs should be shown in your console from which you are running the app.

If that isn't the case, please try running the app from xcode and you will be able to check logs from xcode console.

**Comment by andreasgangso** [comment]
For logs while debugging it's also helpful to set loglevel to verbose when initializing:
```dart
builder.withLogLevel(RudderLogger.VERBOSE);
```

**Comment by chandrabezzo** [comment]
@desusai7 that my log using Android use previous version. In iOS, I never see any log.

**Comment by chandrabezzo** [comment]
@andreasgangso Set log level to VERBOSE but never see anything

**Comment by chandrabezzo** [comment]
@desusai7 maybe contains different code between github master from you and [docker](https://hub.docker.com/r/rudderlabs/rudder-server/tags)?

**Comment by chandrabezzo** [comment]
After coordination with my colleagues @ariretiarno this issue just fixed when you use these version:
- [rudder_sdk_flutter](https://pub.dev/packages/rudder_sdk_flutter): 2.4.1
- [rudder_server](https://github.com/rudderlabs/rudder-server/releases/tag/v1.12.2): 1.12.2

cc: @desusai7 @andreasgangso

**Comment by chandrabezzo** [comment]
Thank you everyone, if found any error again in the future please tell us.

**Comment by chandrabezzo** [comment]
FYI guys, if you use rudder_server version 1.12.2 you can use rudder_sdk_flutter latest version (2.4.1) or
```yaml
  rudder_plugin_android: 2.2.2
  rudder_plugin_ios: 2.2.2
  rudder_plugin_web: 2.2.3
  rudder_sdk_flutter_platform_interface: 2.3.1
```
